strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -0,0 +1,833 @@
+        if self.active[n] is None:
+            raise ValueError("Cannot apply homodyne measurement, mode does not exist")
+        self.phase_shift(phi, n)
+        covmat = self.hbar * np.diag(np.array([eps ** 2, 1.0 / eps ** 2])) / 2
+        indices = [n]
+        vals = np.array([val, 0])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def post_select_heterodyne(self, n, alpha_val):
+        r"""Simulates a heterodyne measurement on a mode, postselecting on an outcome.
+
+        Args:
+            n (int): mode to be measured
+            alpha_val (array): measurement value to post-select
+        """
+        if self.active[n] is None:
+            raise ValueError("Cannot apply heterodyne measurement, mode does not exist")
+
+        covmat = self.hbar * np.identity(2) / 2
+        indices = [n]
+        vals = np.array([alpha_val.real, alpha_val.imag])
+        self.post_select_generaldyne(covmat, indices, vals)
+
+    def apply_u(self, U):
+        r"""Transforms the state according to the linear optical unitary that
+        maps a[i] \to U[i, j]^*a[j].
+
+        Args:
+            U (array): linear opical unitary matrix
+        """
+        Us = symp.interferometer(U)
+        self.means = update_means(self.means, Us, self.from_xp)
+        self.covs = update_covs(self.covs, Us, self.from_xp)
+
+    def apply_channel(self, X, Y):
+        r"""Transforms the state according to a deterministic Gaussian CPTP map.
+
+        Args:
+            X (array): matrix for mutltiplicative part of transformation
+            Y (array): matrix for additive part of transformation.
+        """
+        self.means = update_means(self.means, X, self.from_xp)
+        self.covs = update_covs(self.covs, X, self.from_xp, Y)
+
+    def expandS(self, modes, S):
+        """Expands symplectic matrix on subset of modes to symplectic matrix for the whole system.
+
+        Args:
+            modes (list): list of modes on which S acts
+            S (array): symplectic matrix
+        """
+        return symp.expand(S, modes, self.nlen)
+
+    def expandXY(self, modes, X, Y):
+        """Expands deterministic Gaussian CPTP matrices ``(X,Y)`` on subset of modes to
+        transformations for the whole system.
+
+        Args:
+            modes (list): list of modes on which ``(X,Y)``
+            X (array): matrix for mutltiplicative part of transformation
+            Y (array): matrix for additive part of transformation.
+        """
+        X2 = symp.expand(X, modes, self.nlen)
+        M = len(Y) // 2
+        Y2 = np.zeros((2 * self.nlen, 2 * self.nlen), dtype=Y.dtype)
+        w = np.array(modes)
+
+        Y2[w.reshape(-1, 1), w.reshape(1, -1)] = Y[:M, :M].copy()
+        Y2[(w + self.nlen).reshape(-1, 1), (w + self.nlen).reshape(1, -1)] = Y[M:, M:].copy()
+        Y2[w.reshape(-1, 1), (w + self.nlen).reshape(1, -1)] = Y[:M, M:].copy()
+        Y2[(w + self.nlen).reshape(-1, 1), w.reshape(1, -1)] = Y[M:, :M].copy()
```

Comment:
Might be worth storing some of the repeating reshaped weights as separate objects

Reply:
I think we can just go with Nico's suggestion to do it another way then it will be less confusing to follow.

Reply:
Done.